pkg/controller/externalnode/controller_test.go

```diff
@@ -0,0 +1,606 @@
+				},
+			},
+			expectedEntity: &v1alpha2.ExternalEntity{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      "vm3",
+					Namespace: "ns1",
+					OwnerReferences: []metav1.OwnerReference{
+						{
+							APIVersion: "crd.antrea.io/v1alpha1",
+							Kind:       "ExternalNode",
+							Name:       "vm3",
+						},
+					},
+					Labels: map[string]string{"en": "vm3", "app": "db"},
+				},
+				Spec: v1alpha2.ExternalEntitySpec{
+					Endpoints: []v1alpha2.Endpoint{
+						{IP: "1.1.1.4"},
+					},
+					ExternalNode: "vm3",
+				},
+			},
+		},
+	} {
+		t.Run(tc.name, func(t *testing.T) {
+			controller := newExternalNodeController([]runtime.Object{tc.externalNode})
+			stopCh := make(chan struct{})
+			defer close(stopCh)
+			informerFactory.Start(stopCh)
+			go controller.Run(stopCh)
+			err := wait.PollImmediate(time.Millisecond*50, time.Second, func() (done bool, err error) {
+				entities, listErr := controller.crdClient.CrdV1alpha2().ExternalEntities(tc.externalNode.Namespace).List(context.TODO(), metav1.ListOptions{})
+				if listErr != nil {
+					return false, listErr
+				}
+				if len(entities.Items) == 0 {
+					return false, nil
+				}
+				tempEE := entities.Items[0]
+				if tc.existingEntity == nil {
+					deleteErr := controller.crdClient.CrdV1alpha2().ExternalEntities(tc.externalNode.Namespace).Delete(context.TODO(), tempEE.Name, metav1.DeleteOptions{})
+					if deleteErr != nil {
+						return false, nil
+					}
+				}
+				return true, nil
+			})
+			require.NoError(t, err)
+
+			_, err = controller.crdClient.CrdV1alpha1().ExternalNodes(tc.externalNode.Namespace).Update(context.TODO(), tc.updatedExternalNode, metav1.UpdateOptions{})
+			require.NoError(t, err)
+			err = wait.PollImmediate(time.Millisecond*50, time.Second, func() (done bool, err error) {
+				return checkExternalEntityExists(controller.crdClient, tc.expectedEntity)
+			})
+			assert.NoError(t, err)
+			if tc.existingEntity != nil {
+				exists, err := checkExternalEntityExists(controller.crdClient, tc.existingEntity)
+				assert.NoError(t, err)
+				assert.False(t, exists)
+			}
+		})
+
+	}
+}
+
+func TestDeleteExternalNode(t *testing.T) {
+	externalNode := &v1alpha1.ExternalNode{
+		ObjectMeta: metav1.ObjectMeta{Name: "vm1", Namespace: "ns1", Labels: map[string]string{"en": "vm1"}},
+		Spec: v1alpha1.ExternalNodeSpec{
+			Interfaces: []v1alpha1.NetworkInterface{{IPs: []string{"1.1.1.2"}}},
+		},
+	}
+	expectedEntity := &v1alpha2.ExternalEntity{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "vm1",
+			Namespace: "ns1",
+			OwnerReferences: []metav1.OwnerReference{
+				{
+					APIVersion: "crd.antrea.io/v1alpha1",
+					Kind:       "ExternalNode",
+					Name:       "vm1",
+				},
+			},
+			Labels: map[string]string{"en": "vm1"},
+		},
+		Spec: v1alpha2.ExternalEntitySpec{
+			Endpoints: []v1alpha2.Endpoint{
+				{IP: "1.1.1.2"},
+			},
+			ExternalNode: "vm1",
+		},
+	}
+	testDeleteExternalNode(t, externalNode, expectedEntity)
```

Comment:
same as `testAddExternalNode`, separting input and test code makes the test less readable.

Reply:
updated.